emhttp/plugins/dynamix/include/UpdateTwo.php

```diff
@@ -13,166 +13,254 @@
-          for ($c = 1; $c < count($cmd); $c++) if (scan($cmd[$c],'isolcpus')) {$make |= ($cmd[$c]!=$isolcpus); $cmd[$c] = $isolcpus; break;}
-          // or insert a new setting
-          if ($c==count($cmd) && $isolcpus) {array_splice($cmd,-1,0,$isolcpus); $make = true;}
-          $syslinux[$n] = '  '.str_replace('  ',' ',implode(' ',$cmd));
-        }
-        $n++;
-      }
-      $i = $n - 1;
-    }
-    $i++;
-  }
-  if ($make) file_put_contents($cfg, implode("\n",$syslinux)."\n");
-  $reply = ['success' => $name];
-  break;
+	/* Path to syslinux configuration file */
+	$cfg = '/boot/syslinux/syslinux.cfg';
+
+	/* Read the syslinux configuration file into an array, ignoring empty lines */
+	$syslinux = file($cfg, FILE_IGNORE_NEW_LINES + FILE_SKIP_EMPTY_LINES);
+	$size = count($syslinux);
+	$make = false;
+
+	/* Path to the temporary file containing new isolcpus settings */
+	$file = "/var/tmp/$name.tmp";
+	$isolcpus = file_get_contents($file);
+
+	if ($isolcpus != '') {
+		/* Convert isolcpus string to an array of numbers and sort them */
+		$numbers = explode(',', $isolcpus);
+		sort($numbers, SORT_NUMERIC);
+
+		/* Initialize variables for range conversion */
+		$isolcpus = $previous = array_shift($numbers);
+		$range = false;
+
+		/* Convert sequential numbers to a range */
+		foreach ($numbers as $number) {
+			if ($number == $previous + 1) {
+				$range = true;
+			} else {
+				if ($range) {
+					$isolcpus .= '-' . $previous;
+					$range = false;
+				}
+				$isolcpus .= ',' . $number;
+			}
+			$previous = $number;
+		}
+		if ($range) {
+			$isolcpus .= '-' . $previous;
+		}
+
+		/* Format isolcpus string for syslinux configuration */
+		$isolcpus = "isolcpus=$isolcpus";
+	}
+
+	/* Remove the temporary file */
+	unlink($file);
+
+	$i = 0;
+	while ($i < $size) {
+		/* Find sections in syslinux config and exclude safemode */
+		if (scan($syslinux[$i], 'label ') && !scan($syslinux[$i], 'safe mode') && !scan($syslinux[$i], 'safemode')) {
+			$n = $i + 1;
+
+			/* Find the current requested setting */
+			while ($n < $size && !scan($syslinux[$n], 'label ')) {
+				if (scan($syslinux[$n], 'append ')) {
+					$cmd = preg_split('/\s+/', trim($syslinux[$n]));
+
+					/* Replace an existing isolcpus setting */
+					for ($c = 1; $c < count($cmd); $c++) {
+						if (scan($cmd[$c], 'isolcpus')) {
+							$make |= ($cmd[$c] != $isolcpus);
+							$cmd[$c] = $isolcpus;
+							break;
+						}
+					}
+
+					/* Or insert a new isolcpus setting if not found */
+					if ($c == count($cmd) && $isolcpus) {
+						array_splice($cmd, -1, 0, $isolcpus);
+						$make = true;
+					}
+
+					/* Update the syslinux configuration line */
+					$syslinux[$n] = '  ' . str_replace('  ', ' ', implode(' ', $cmd));
+				}
+				$n++;
+			}
+			$i = $n - 1;
+		}
+		$i++;
+	}
+
+	/* Write the updated syslinux configuration back to the file if changes were made */
+	if ($make) {
+		file_put_contents($cfg, implode("\n", $syslinux) . "\n");
```

Comment:
Please implement the new `file_put_contents_atomic()` function (from Wrappers.php) in place of `file_put_contents()` in this PR. Implementing that everywhere will be a separate project, but might as well implement it here now since you are working on this file.

Reply:
Resolved.